src/DB/ProductMetaQueryHelper.php

```diff
@@ -0,0 +1,135 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\DB;
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Exception\InvalidMeta;
+use Automattic\WooCommerce\GoogleListingsAndAds\Infrastructure\Service;
+use Automattic\WooCommerce\GoogleListingsAndAds\PluginHelper;
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductMetaHandler;
+use wpdb;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class ProductMetaQueryHelper.
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\DB
+ */
+class ProductMetaQueryHelper implements Service {
+
+	use PluginHelper;
+
+	protected const BATCH_SIZE = 500;
+
+	/**
+	 * @var wpdb
+	 */
+	protected $wpdb;
+
+	/**
+	 * ProductMetaQueryHelper constructor.
+	 *
+	 * @param wpdb $wpdb
+	 */
+	public function __construct( wpdb $wpdb ) {
+		$this->wpdb = $wpdb;
+	}
+
+	/**
+	 * Get all values for a given meta_key as post_id=>meta_value.
+	 *
+	 * @param string $meta_key The meta value to retrieve for all posts.
+	 * @return array Meta values by post ID.
+	 *
+	 * @throws InvalidMeta If the meta key isn't valid.
+	 */
+	public function get_all_values( string $meta_key ): array {
+		self::validate_meta_key( $meta_key );
+
+		$query = "SELECT post_id, meta_value FROM {$this->wpdb->postmeta} WHERE meta_key = %s";
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+		$results = $this->wpdb->get_results( $this->wpdb->prepare( $query, $this->prefix_meta_key( $meta_key ) ) );
+		$return  = [];
+		foreach ( $results as $r ) {
+			$return[ $r->post_id ] = $r->meta_value;
+		}
+		return $return;
+	}
+
+	/**
+	 * Insert a meta value for multiple posts.
+	 *
+	 * @param string $meta_key The meta value to insert.
+	 * @param array  $statuses Array of [post_id=>status,…].
+	 *
+	 * @throws InvalidMeta If the meta key isn't valid.
+	 */
+	public function batch_insert_values( string $meta_key, array $statuses ) {
+		if ( empty( $statuses ) ) {
+			return;
+		}
+
+		self::validate_meta_key( $meta_key );
+		$meta_key = $this->prefix_meta_key( $meta_key );
+
+		foreach ( array_chunk( $statuses, self::BATCH_SIZE, true ) as $statuses_chunk ) {
+			$values        = [];
+			$place_holders = [];
+			foreach ( $statuses_chunk as $post_id => $status ) {
+				$place_holders[] = '(%s, %s, %s)';
+				$values[]        = $meta_key;
+				$values[]        = $post_id;
+				$values[]        = $status;
+			}
+			$query  = "INSERT INTO {$this->wpdb->postmeta} (`meta_key`, `post_id`, `meta_value` ) VALUES ";
+			$query .= implode( ', ', $place_holders );
+			// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+			$this->wpdb->query( $this->wpdb->prepare( $query, $values ) );
```

Comment:
Do we need to "albeit" silently log any errors here?

Reply:
Debug log added: 595e9abb08a55c7f343f13a4118f2fe2370bfd8e